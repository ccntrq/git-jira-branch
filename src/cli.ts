import {ValidationError} from '@effect/cli';
import type {CliApp} from '@effect/cli/CliApp';
import * as Command from '@effect/cli/Command';
import * as HelpDoc from '@effect/cli/HelpDoc';
import * as Span from '@effect/cli/HelpDoc/Span';
import type * as CommandExecutor from '@effect/platform/CommandExecutor';
import {Console, Effect, pipe} from 'effect';

import * as packageJson from '../package.json';
import {create} from './commands/create/create.command';
import {deleteCommand} from './commands/delete/delete.command';
import {info} from './commands/info/info.command';
import {list} from './commands/list/list.command';
import {open} from './commands/open/open.command';
import {switchCommand} from './commands/switch/switch.command';
import {tidy} from './commands/tidy/tidy.command';
import type {NoAssociatedBranch} from './schema/no-associated-branch';
import type {AppConfigService} from './services/app-config';
import type {GitClient} from './services/git-client';
import type {JiraClient} from './services/jira-client';
import type {GitJiraBranchError} from './types';

// for version and help
const gitJiraBranch = pipe(Command.make('git-jira-branch', {}));

const mainCommand = gitJiraBranch.pipe(
  Command.withSubcommands([
    create,
    switchCommand,
    deleteCommand,
    open,
    info,
    list,
    tidy,
  ]),
);

const cli = {
  name: packageJson.name,
  version: packageJson.version,
  summary: Span.text(packageJson.description),
};

export const cliEffect = (
  args: Array<string>,
): Effect.Effect<
  void,
  GitJiraBranchError | ValidationError.ValidationError | NoAssociatedBranch,
  | CliApp.Environment
  | GitClient
  | AppConfigService
  | JiraClient
  | CommandExecutor.CommandExecutor
> =>
  Command.run(
    mainCommand,
    cli,
  )(args).pipe(
    Effect.tapError((e) => {
      if (ValidationError.isValidationError(e)) {
        // handled and printed by the cli library already
        return Effect.void;
      }
      return printErrors(HelpDoc.p(Span.error(e.message)));
    }),
  );

const printErrors = (doc: HelpDoc.HelpDoc): Effect.Effect<void> =>
  Console.error(HelpDoc.toAnsiText(doc));
