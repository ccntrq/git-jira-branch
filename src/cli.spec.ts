import {live} from '@effect/vitest';
import {Effect, pipe} from 'effect';

import {cliEffect} from './cli';

import {afterEach, describe, expect, vi} from 'vitest';
import {cliTestLayer} from './test/mock-implementations';

const mockLog = vi.spyOn(console, 'log').mockImplementation(() => undefined);

const withBaseArgs = (args: Array<string>): Effect.Effect<Array<string>> =>
  Effect.sync(() => ['node', 'git-jira-branch', ...args]);

describe('cli', () => {
  describe('main command', () => {
    afterEach(() => {
      vi.clearAllMocks();
    });

    live('should print version (--version)', () =>
      Effect.gen(function* () {
        yield* Effect.provide(
          pipe(withBaseArgs(['--version']), Effect.flatMap(cliEffect)),
          cliTestLayer,
        );

        expect(mockLog.mock.calls[0]?.[0]).toMatch(/\d+\.\d+\.\d+/);
      }),
    );

    live('should print help (--help)', () =>
      Effect.gen(function* () {
        yield* Effect.provide(
          pipe(withBaseArgs(['--help']), Effect.flatMap(cliEffect)),
          cliTestLayer,
        );

        expect(
          (mockLog.mock.calls[0]?.[0] as string)
            .split(/\n/)
            .slice(3)
            .join('\n'),
        ).toMatchSnapshot();
      }),
    );
  });
});
