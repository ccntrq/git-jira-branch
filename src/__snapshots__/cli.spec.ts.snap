// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cli > main command > should print help (--help) 1`] = `
"
[0;1mUSAGE[0m

$ git-jira-branch

[0;1mOPTIONS[0m

[0;1m--completions sh | bash | fish | zsh[0m

  One of the following: sh, bash, fish, zsh

  Generate a completion script for a specific shell.

  This setting is optional.

[0;1m--log-level all | trace | debug | info | warning | error | fatal | none[0m

  One of the following: all, trace, debug, info, warning, error, fatal, none

  Sets the minimum log level for a command.

  This setting is optional.

[0;1m(-h, --help)[0m

  A true or false value.

  Show the help documentation for a command.

  This setting is optional.

[0;1m--wizard[0m

  A true or false value.

  Start wizard mode for a command.

  This setting is optional.

[0;1m--version[0m

  A true or false value.

  Show the version of the application.

  This setting is optional.

[0;1mCOMMANDS[0m

  - create [(-b, --base text)] [(-t, --type text)] [(-r, --reset)] <jira-key>  
Fetches the given Jira ticket and creates an aproriately named branch for it.
The branch type (bug or feat) is determined by the ticket type. The branch name
is based on the ticket summary.

  - switch <jira-key>                                                          
Switches to an already existing branch that is associated with the given Jira
ticket.

  - delete [--force] <jira-key>                                                
Deletes the branch associated with the given Jira Ticket.

  - open [<jira-key>]                                                          
Opens the given Jira ticket in your default browser. If no ticket is given the
jira ticket for the current branch is opened.

  - info [<jira-key>]                                                          
Displays information for the given Jira ticket on your terminal. If no ticket is
provided, it presents information for the Jira ticket associated with the
current branch.

  - list                                                                       
Lists all branches that appear to be associated with a Jira ticket.

  - tidy [--force]                                                             
Deletes branches for tickets that are done.
"
`;
